pkg/cloudprovider/provider/otc/provider.go

```diff
@@ -0,0 +1,942 @@
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+	providerconfigtypes "github.com/kubermatic/machine-controller/pkg/providerconfig/types"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+)
+
+const (
+	floatingIPReleaseFinalizer = "kubermatic.io/release-otc-floating-ip"
+	floatingIPIDAnnotationKey  = "kubermatic.io/release-otc-floating-ip"
+	cloudName                  = "otc.t-systems.com"
+)
+
+// clientGetterFunc returns an OpenStack client.
+type clientGetterFunc func(c *Config) (*gophercloud.ProviderClient, error)
+
+// serverReadinessWaiterFunc waits for the server with the given ID to be
+// ACTIVE.
+type serverReadinessWaiterFunc func(computeClient *gophercloud.ServiceClient, serverID string) error
+
+type provider struct {
+	configVarResolver     *providerconfig.ConfigVarResolver
+	clientGetter          clientGetterFunc
+	serverReadinessWaiter serverReadinessWaiterFunc
+}
+
+// New returns a OTC provider
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloudprovidertypes.Provider {
+	return &provider{
+		configVarResolver:     configVarResolver,
+		clientGetter:          getClient,
+		serverReadinessWaiter: waitUntilInstanceIsActive,
+	}
+}
+
+type Config struct {
+	AccessKey  string
+	SecretKey  string
+	DomainName string
+	TenantID   string
+	TokenID    string
+	VPCID      string
+	Region     string
+
+	// Machine details
+	Image                 string
+	Flavor                string
+	SecurityGroups        []string
+	Network               string
+	Subnet                string
+	FloatingIPPool        string
+	AvailabilityZone      string
+	TrustDevicePath       bool
+	RootDiskSizeGB        *int
+	NodeVolumeAttachLimit *uint
+
+	Tags map[string]string
+}
+
+func (c *Config) Endpoint(kind string) string {
+	return fmt.Sprintf("https://%s.%s.%s", kind, c.Region, cloudName)
+}
+
+const (
+	machineUIDMetaKey = "machine-uid"
+	securityGroupName = "kubernetes-v1"
+
+	instanceReadyCheckPeriod  = 2 * time.Second
+	instanceReadyCheckTimeout = 2 * time.Minute
+)
+
+// Protects floating ip assignment
+var floatingIPAssignLock = &sync.Mutex{}
+
+func (p *provider) getConfig(s v1alpha1.ProviderSpec) (*Config, *providerconfigtypes.Config, *otctypes.RawConfig, error) {
+	if s.Value == nil {
+		return nil, nil, nil, errors.New("machine.spec.providerconfig.value is nil")
+	}
+
+	pconfig := providerconfigtypes.Config{}
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var rawConfig otctypes.RawConfig
+	err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	c := Config{}
+	c.AccessKey, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.AccessKey, "OTC_ACCESS_KEY")
```

Comment:
Those names, `OTC_*`, where do they come from? I'm asking because for example in [terraform](https://www.terraform.io/docs/providers/opentelekomcloud/index.html#configuration-reference) they use standard openstack names

Reply:
I came up with them, thinking it's sensible to have separate names from Openstack, so that you can have an Openstack and an OTC preset in the same container.

Reply:
My problem with custom (diverged from the upstream) names, is that later we going to need to ask users of kubeone to copy their terraform OS_* to OTC_*, or even worse, to convert them manually (it's worse because it's implicit and there will be no way to prevent it). So I'd prefer very much the way it's represented in terraform. Not because it's right, but because it's established.

Reply:
I have no strong feelings either way, but if we *ever* run into a conflict, I'll reserve the right to say "I tOlD yOu So!" ;-)

Reply:
I agree with @kron4eg that we should stick with commonly used variables. If they use `OS_*` variables in their Terraform provider, I think that we should use the same variables in machine-controller. If we want to be sure, we can check what they use in other projects (e.g. CCM or any other project they have), and decide based on that.

> that you can have an Openstack and an OTC preset in the same container.

Do we have some idea when we might want to do that?

Reply:
I could ask the same question: Why did they prefix their access key with "OS_"? (OS_ACCESS_KEY)? Do they really think that there might be other access keys? They obviously did it to be careful and a good citizen and not clobber the program environment.

We want to treat OTC as its own provider and not just an openstack flavor. So it IMHO made and still makes sense to have variables specific to that provider. IMHO it's a terrible choice from Terraform to re-use "OS_" for the OTC provider, but alas, that ship has sailed.

I don't mind the name for the env variables either way, but I find it funny that I suddenly have to defend assining project-specific prefixes to avoid naming conflicts :-D Either OTC is 100% compatible, then we can re-use the Openstack stuff, or we actually treat it as its own thing.

> Do we have some idea when we might want to do that?

Have an e2e test that uses both the openstack and OTC presets in Prow, because you want to test in multiple clouds. Boom. Have the same .direnv in your terraform directory that manages nodes in multiple clusters. Boom. Is it in our multi-cluster multi-cloud world really so hard to imagine that one would for whatever reason have an environment for multiple cloud providers? :-D

Reply:
Changed to OS_.

Reply:
Need to change the preset, so the tests fill fail for now.

Reply:
Done and done.